circuit/gas_constraints.go

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright © 2022 ZkBNB Protocol
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package circuit
+
+import (
+	"errors"
+	"log"
+
+	"github.com/bnb-chain/zkbnb-crypto/circuit/types"
+	"github.com/consensys/gnark/std/signature/eddsa"
+)
+
+type GasAccountConstraints struct {
+	AccountIndex    Variable
+	AccountNameHash Variable
+	AccountPk       eddsa.PublicKey
+	Nonce           Variable
+	CollectionNonce Variable
+	AssetRoot       Variable
+	AssetsInfo      []types.AccountAssetConstraints
+	GasAssetCount   int
+}
+
+type GasConstraints struct {
+	GasAssetCount                   int
+	AccountInfoBefore               GasAccountConstraints
+	MerkleProofsAccountBefore       [AccountMerkleLevels]Variable
+	MerkleProofsAccountAssetsBefore [][AssetMerkleLevels]Variable
+}
+
+func VerifyGas(
+	api API,
+	gas GasConstraints,
+	needGas Variable,
+	gasAssetDeltas []Variable,
+	hFunc MiMC,
+	accountRoot Variable) (newAccountRoot Variable, err error) {
+	newAccountRoot = accountRoot
+	newAccountAssetsRoot := gas.AccountInfoBefore.AssetRoot
+
+	types.IsVariableEqual(api, needGas, gas.AccountInfoBefore.AccountIndex, types.GasAccountIndex)
```

Comment:
If `GasAccountIndex` is a constant, anyway we don't verify it, but always use it?

Reply:
changed